File: GoussasAlexander/HaskellSolutions/lib/Aoc/Day/Five/PartOne.hs
module Aoc.Day.Five.PartOne where

import           Aoc.Class
import           Aoc.Day.Five.Common
import           Aoc.Util


import           Data.List           (foldl')
import           Data.Proxy
import           Data.Vector         (Vector, (!), (//))

data PartOne

instance Aoc (Proxy PartOne) where
  type Result (Proxy PartOne) = [Char]

  solution _ input = do
    let (crates, moves) = runParser parser input
        stacks = mkStacks crates
        stacks' = foldl' processMove stacks moves
     in pure $ getTopsOfStacks stacks'

updateStacks :: Vector [Char] -> Move -> Vector [Char]
updateStacks vs (Move _ from to) =
    vs // [(from, tail $ vs ! from), (to, head (vs ! from) :  vs ! to)]

processMove :: Vector [Char] -> Move -> Vector [Char]
processMove vs move@(Move quantity _ _) =
  foldl' (\vs' _ -> updateStacks vs' move) vs [1..quantity]

File: CoteraWashington/src/main/java/io/github/wgcotera/aoc/day_05/PartOne.java
package io.github.wgcotera.aoc.day_05;

import io.github.wgcotera.aoc.Aoc;

import java.util.List;

import static io.github.wgcotera.aoc.day_05.Common.*;

public class PartOne implements Aoc<String> {
    public static void main(String[] args) {
        System.out.println(new PartOne().run());
    }

    private static List<Stack> getFinalStacks(String input) {
        List<Stack> stacks = listOfStacks(input);
        List<Move> listOfMoves = listOfMoves(input);

        for (Move move : listOfMoves) {

            Stack from = stacks.get(move.from - 1);
            Stack to = stacks.get(move.to - 1);

            to.addElementToStack(from.removeCratesFromStackOneByOne(move.quantity));

        }
        return stacks;
    }

    @Override
    public String solution(String input) {
        return createStringWithTopCrates(getFinalStacks(input), input);
    }

    @Override
    public String inputFilePath() {
        return "5";
    }
}
